### ISSUE TITLE
RFC: `?` operator

### ISSUE THREAD

**Issue by ryzhyk** [vmware/differential-datalog#707]
I am thinking of adding syntactic sugar to work with `Result<>` and `Option<>` types.

DDlog does not have NULLs (NULLs are evil!).  Instead missing values are represented in a type-safe fashion using `Option` and `Result` types. The programmer must open up these types using pattern matching or unwrap them using convenience functions like `unwrap_or` and `unwrap_or_default` in order to access their inner values.

The downside of this is that the code gets bloated, especially when working with many, potentially nested, optional values.  Rust has the `?` operator that makes error handling extremely ergonomic: it unwraps a `Result`, returning from the current function on error.  Groovy has something similar, but the semantics of Groovy's `?` operator is much more confusing.

So what should the DDlog `?` operator do? I suggest an approach similar to Rust:
- `?` applies to expressions of type `Option` and `Result` only.
- if the value is `Some` or `Ok`, it just returns the inner value
- otherwise (`None` or `Err`), terminates the current function, returning a value according to the following table. 

| expression type | expression value | function return type |     return value     |
|:---------------:|:----------------:|:--------------------:|:--------------------:|
| Option< X >       | None             | T                    | default value of T   |
| Result<X,E>     | Err{e}           | Result<_,E>          | Err{e}               |
| Result<X,E>     | Err{e}           | Result<_,F>          | rejected by compiler |
| Result<X,E>     | Err{e}           | any other type T     | default value of T   |

Example
--------

```
function foo(): bool
{
   ...
    var x = match (y) {
        None -> return false,
        Some{z} -> {
             z+5
        }
    };
    ...
}
```

becomes

```
function foo(): bool
{
   ...
   var x = y? + 5;
   ...
}
```

One open question is whether `?` should be allowed outside of a function, in particular inside a rule:
- in a relation literal
- in an assignment clause
- in a filter clause

In all these cases, the behavior of `?` would still be well defined (just replace the "function return type" with expression type expected in the given context in the above table), but it may be less intuitive to the user.

@mbudiu-vmw , @blp

**Comment by mbudiu-vmw** [comment]
I think it's dangerous to allow it everywhere. I would suggest by starting to only allow it in functions that already return Option or Result, there the meaning is very clear. If this is not enough to make the code readable we can talk again.
But people have to think how to handle these errors, just returning default(T) will prevent them from thinking.

**Comment by ryzhyk** [comment]
@mbudiu-vmw , I think I agree.  It is not completely clear how useful the limited feature will be, but it is a good idea to implement it and see how it affects code that people will write.

**Comment by blp** [comment]
I think that Rust defines these things in terms of traits, which presumably allows an extra level of flexibility. DDlog doesn't have traits and could not have the same extra flexibility in the same way. I don't know whether that's actually important, because I haven't written large-scale Rust code.

**Comment by blp** [comment]
The "default value of T" cases seem really risky to me.

I like the others.

**Comment by blp** [comment]
I really like the Rust ? operator. I often wish for it in other languages.

**Comment by ryzhyk** [comment]
> I think that Rust defines these things in terms of traits, which presumably allows an extra level of flexibility. DDlog doesn't have traits and could not have the same extra flexibility in the same way. I don't know whether that's actually important, because I haven't written large-scale Rust code.

True, we do lose some flexibility that traits would otherwise enable.   But in my experience, traits don't play a huge role in error handling.  First, `?` only applies to `Result` and `Option` types.  In case of `Result`, traits are potentially useful to convert the error type of the expression to the error type of the function, but more often than not I find that either the two types are the same (in which case traits are not needed) or they are different and I need to do the conversion manually, because there is not `From` trait to convert between the two.

**Comment by ryzhyk** [comment]
> The "default value of T" cases seem really risky to me.

I agree, it does make it easy to shoot yourself in the foot. I'll restrict the implementation to `Option` and `Result` types only.

**Comment by ryzhyk** [comment]
Implemented in #710
